sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -453,14 +459,78 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
             DoFn.OutputReceiver.class.getSimpleName(),
             formatType(paramT),
             formatType(expectedOutputReceiverT));
-        extraParameters.add(DoFnSignature.Parameter.OUTPUT_RECEIVER);
+        extraParameters.add(DoFnSignature.Parameter.outputReceiver());
       } else if (RestrictionTracker.class.isAssignableFrom(rawType)) {
         errors.checkArgument(
-            !extraParameters.contains(DoFnSignature.Parameter.RESTRICTION_TRACKER),
+            !extraParameters.contains(DoFnSignature.Parameter.restrictionTracker()),
             "Multiple %s parameters",
             RestrictionTracker.class.getSimpleName());
-        extraParameters.add(DoFnSignature.Parameter.RESTRICTION_TRACKER);
+        extraParameters.add(DoFnSignature.Parameter.restrictionTracker());
         trackerT = paramT;
+      } else if (State.class.isAssignableFrom(rawType)) {
+        // m.getParameters() is not available until Java 8
+        Annotation[] annotations = m.getParameterAnnotations()[i];
+        String id = null;
+        for (Annotation anno : annotations) {
+          if (anno.annotationType().equals(DoFn.StateId.class)) {
+            id = ((DoFn.StateId) anno).value();
+            break;
+          }
+        }
+        errors.checkArgument(
+            id != null,
+            "%s parameter of type %s at index %s missing %s annotation",
+            fnClass.getRawType().getName(),
+            params[i],
+            i,
+            DoFn.StateId.class.getSimpleName());
+
+        errors.checkArgument(
+            !stateParameters.containsKey(id),
+            "%s parameter of type %s at index %s duplicates %s(\"%s\") on other parameter",
+            fnClass.getRawType().getName(),
+            params[i],
+            i,
+            DoFn.StateId.class.getSimpleName(),
+            id);
+
+        // By static typing this is already a well-formed State subclass
+        TypeDescriptor<? extends State> stateType =
+            (TypeDescriptor<? extends State>)
+                TypeDescriptor.of(fnClass.getType())
+                    .resolveType(params[i]);
+
+        StateDeclaration stateDecl = stateDeclarations.get(id);
+        errors.checkArgument(
+            stateDecl != null,
+            "%s parameter of type %s at index %s references undeclared StateId \"%s\"",
```

Comment:
Why is StateId referenced here explicitly, but as `DoFn.StateId.class.getSimpleName()` above?